### DISCUSSION TITLE
How can I add the Identity.UI to my OrchardCore based application?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#6960] [Q&A] [answered]
I am trying to use `OrchardCore` to add multi-tenants/modular support to an asp.net core 3.1 application without the "CMS" functions.

My application uses `Identity` for authentication. I want to use the [Identity.UI](https://www.nuget.org/packages/Microsoft.AspNetCore.Identity.UI) to have a complete UI around identity.

I have the following 3 projects.
 1. Application.Web
 2. Application.Modules.Module1
 3. Application.Modules.Module2

The `Application.Web` is the main application and I installed `OrchardCore.Application.Mvc.Targets`. Also it has a reference to `Application.Modules.Module1` and `Application.Modules.Module2`.

When I run my application everything is working as expected. But now, I need to add the `Identity.UI` to give the user a UI to be able to login, register and all other common identity functions.

I added `Microsoft.AspNetCore.Identity.UI` to my `Application.Web` project which added `Areas/Identity`  to my solution. Then I configured the Identity from `Application.Web.Startup.ConfigureServices(...)` like this

```
public void ConfigureServices(IServiceCollection services)
{
    services.AddOrchardCore()
            .AddMvc()
            .WithTenants()
            .ConfigureServices((tenantServices, serviceProvider) =>
            {
                // we create an instance of the configuration
                IShellConfiguration tenantConfigs = serviceProvider.GetRequiredService<IShellConfiguration>();

                tenantServices.AddRazorPages();
                tenantServices.AddDbContext<ApplicationContext>(options =>
                {
                    string connectionString = tenantConfigs.GetConnectionString("DefaultConnection");

                    options.UseSqlServer(connectionString);
                });

                tenantServices.AddIdentity<User, UserRole>(options =>
                {
                    // TODO: this would be read from the tenantConfigs
                    options.SignIn.RequireConfirmedAccount = true;
                    options.SignIn.RequireConfirmedEmail = true;
                    options.Lockout.DefaultLockoutTimeSpan = TimeSpan.FromMinutes(10);
                }).AddEntityFrameworkStores<ApplicationContext>()
                  .AddDefaultTokenProviders();

                tenantServices.ConfigureApplicationCookie(options =>
                {
                    options.LoginPath = $"/Identity/Account/Login";
                    options.LogoutPath = $"/Identity/Account/Logout";
                    options.AccessDeniedPath = $"/Identity/Account/AccessDenied";

                    options.ExpireTimeSpan = TimeSpan.FromMinutes(15);
                    options.SlidingExpiration = true;
                    options.Cookie.HttpOnly = true;
                });
            });
}
```

However, non of the identity route is working. I get white page when I got to `/Identity/Account/Login`. What is the trick into registering the `Identity.UI` correctly with OrchardCore?

**Comment by ns8482e** [comment, marked as answer]
Razor library is not supported See. #4247

However you care create your own module that contains Identity scaffolder using `dotnet aspnet-codegenerator identity` and move your all files from `YourModule\Areas\Idenity` to `YourModule`

**Reply by MikeAlhayek** [reply]
@ns8482e So razor pages are supported in the module itself but not in the main application? If that is the case, then how would I configure identity based on the tenant?

Here is how I configure services at the tenant level, but not sure how to do the same thing from the module. 

```
services.AddOrchardCore()
        .AddMvc()
        .WithTenants()
        .ConfigureServices((tenantServices, serviceProvider) =>
        {
            // This is where I would typically add tenant based options
        });
```

**Reply by ns8482e** [reply]
@CrestApps  Razor pages are supported. 
However in OC modules are placed inside virtual area, hence `Area` fodder within module or app is not supported.

And you can use following to configure services in module
```
 services
.AddOrchardCore()
.ConfigureServices()

```

**Comment by MikeAlhayek** [comment]
I finally got it to work. I moved the pages from `~/Areas/Identity/Pages` to `~/Pages` then added `services.AddRazorPages();`
